ddprof-lib/src/main/cpp/threadFilter.cpp

```diff
@@ -17,12 +17,24 @@ ThreadFilter::~ThreadFilter() {
     std::unique_lock<std::mutex> lock(_slot_mutex);
     _slots.clear();
 }
-
 ThreadFilter::SlotID ThreadFilter::registerThread() {
+    int top = _free_list_top.load(std::memory_order_acquire);
+    for (int i = 0; i < top; ++i) {
+        int value = _free_list[i].load(std::memory_order_relaxed);
+        if (value >= 0) {
+            int expected = value;
+            if (_free_list[i].compare_exchange_strong(expected, -1, std::memory_order_acq_rel)) {
+                return value; // Successfully claimed a free slot
+            }
+            // If CAS fails, someone else claimed it, continue scanning
+        }
+    }
+
```

Comment:
This looks good. Have you experimented with a custom stride or thread-local random starting offset to improve the contended situation? Basically, to prevent all contending threads starting at 0 and then competing for all subsequent slots until everyone is satisfied - forcing the most unlucky thread to do N checks, N being the number of competing threads.